crates/sui-upload-display/src/main.rs

```diff
@@ -0,0 +1,576 @@
+                    "Initializing GCS with bucket '{}' using credentials from '{}'",
+                    bucket, cred_path
+                );
+
+                let gcs_builder = GoogleCloudStorageBuilder::new()
+                    .with_service_account_path(cred_path.clone())
+                    .with_bucket_name(bucket.clone());
+
+                match gcs_builder.build() {
+                    Ok(store) => {
+                        info!("Successfully initialized GCS client");
+                        Some(Arc::new(store))
+                    }
+                    Err(e) => {
+                        warn!("Failed to initialize GCS client: {}", e);
+                        None
+                    }
+                }
+            }
+            _ => {
+                warn!(
+                    "Either GCS cred path or bucket is not set, display data will not be uploaded"
+                );
+                None
+            }
+        };
+
+    let registry = Registry::new();
+    let metrics = IndexerMetrics::new(&registry);
+
+    let remote_url = Url::parse(&config.remote_url).context("Failed to parse remote URL")?;
+    let client = IngestionClient::new_remote(remote_url, metrics)?;
+    info!("Initialized remote client with URL: {}", config.remote_url);
+
+    let cancellation_token = CancellationToken::new();
+    let mut accumulated_display_data = None;
+
+    let (starting_checkpoint, latest_file_info) = find_last_processed_checkpoint(&config).await?;
+    if let Some((epoch, file_path)) = latest_file_info {
+        match load_display_entries_from_file(&config, &file_path, epoch).await {
+            Ok(loaded_data) => {
+                info!(
+                    "Successfully loaded display data from previous epoch {} with {} entries",
+                    epoch,
+                    loaded_data.displays.len()
+                );
+                accumulated_display_data = Some(loaded_data);
+            }
+            Err(e) => {
+                warn!("Failed to load display entries from file: {}", e);
+                // Continue without previous data
+            }
+        }
+    }
+
+    info!(
+        "Service initialized. Processing checkpoints concurrently with concurrency limit {} and batch size {}", 
+        config.concurrency_limit,
+        config.batch_size
+    );
+    let display_updates = Arc::new(DisplayUpdateMap::new());
+    let config = Arc::new(config);
+    let client = Arc::new(client);
+
+    let mut checkpoint_processed = starting_checkpoint;
+
+    loop {
+        let end_checkpoint = checkpoint_processed + config.batch_size;
+        let checkpoints: Vec<u64> = (checkpoint_processed..end_checkpoint).collect();
+        let client_ref = client.as_ref();
+        let display_updates_ref = &display_updates;
+
+        let has_errors = Arc::new(AtomicBool::new(false));
+        stream::iter(checkpoints)
+            .for_each_concurrent(config.concurrency_limit, |checkpoint| {
```

Comment:
I think you want to use `IngestionService` rather than `IngestionClient` -- it handles the concurrent etc for you -- you can call `run` on it with an iterator of checkpoint sequence numbers.